### ISSUE TITLE
How to achieve on-demand data loading

### ISSUE THREAD

**Issue by shaofing** [dotnet/efcore#32540]
I would like to implement on-demand data loading functionality in Entity Framework Core, similar to DataReader.Read(). Currently, the default behavior of EF Core is to load the entire result set into memory, rather than loading it row by row. While the design philosophy of EF Core is to retrieve data through LINQ queries and return the results as collections, I would like to be able to load data on demand to reduce memory consumption.

I suggest implementing on-demand data loading functionality to provide more flexibility in controlling the data loading mechanism when using EF Core, such as providing a method similar to DataReader.Read() to load data row by row, or offering a more flexible mechanism for paging data loading.

Thank you!

```C#

            //EF Core
            {
                using (var db = new Wms7Context())
                {
                    var lst = db.receipt_detail.OrderBy(x => x.RECEIPT_ID).ThenBy(x => x.LINE_ID);
                    //When executing foreach, all data is loaded into memory.
                    foreach (var item in lst)
                    {
                        if (item.EXTERNAL_RECEIPT_ID == "str1")
                            break;
                    }
                }
            }
            //Dapper
            {
                // When CommandFlags.None is used, executing foreach will load data on demand.
                var lst = conn.Query<receipt_detail>(new CommandDefinition(sql, flags: CommandFlags.None));
                foreach (var item in lst)
                {
                    if (item.EXTERNAL_RECEIPT_ID == "str1")
                        break;
                }
            }

```

**Comment by roji** [comment]
> Currently, the default behavior of EF Core is to load the entire result set into memory, rather than loading it row by row

That's incorrect - what gave you that impression? Have your seen the documentation on [buffering vs. streaming](https://learn.microsoft.com/en-us/ef/core/performance/efficient-querying#buffering-and-streaming)?

Specifically, the code sample you posted above - where `foreach` is used to enumerate over the results - streams data from the database and does not load it all into memory (assuming you're not using a retrying strategy).

**Comment by roji** [comment]
@shaofing none of this has anything to do with AsNoTracking. It's only a question of which operator you use to execute your LINQ query (foreach streams, ToList does not).

**Comment by shaofing** [comment]
I need to query data exceeding 800,000.
Previously, without using AsNoTracking(), there was continuous memory growth,
but after using AsNoTracking(), memory usage is very low.
My problem description may not be very accurate, but the core issue I need to address is memory growth.
Thank you, this issue can be closed.

**Comment by roji** [comment]
OK, yeah - in a sense, tracking queries also "buffers", since it snapshots every you load, in case you want to change it later.
